DOC add priorities update 09/2024 (#1)
glemaitre/probabl-blog-post: 09_2024_scikit_learn_priorities.md

Let's review our achievements and set the stage for our upcoming focus areas.

## Retrospective on the past 6 months

This section summarizes the progress of Probabl's work on the different supported
open-source projects.

### [`scikit-learn`](https://github.com/scikit-learn/scikit-learn)

In a previous [blog
post](https://papers.probabl.ai/scikit-learns-priorities-at-probabl), we outlined
Probabl's priorities for `scikit-learn`. In this section, we report on the progress of
these priorities.

#### Short-term priorities

We successfully completed two major short-term priorities:

1. Website Migration: We successfully migrated the `scikit-learn` website to the new
   PyData theme-based website. This modernization effort improves the user
   experience and aligns our documentation with the broader PyData ecosystem.
1. SLEP6 - Metadata Routing: We completed adapting most estimators to handle
   metadata routing as specified in
   [SLEP6](https://scikit-learn-enhancement-proposals.readthedocs.io/en/latest/slep006/proposal.html).
   This new API consistently handles metadata (e.g., `sample_weight`, `groups`, etc.)

>    for all estimators. It allows for fewer bugs and enables new

Reviewer: ```suggestion
   for almost all estimators. It allows for fewer bugs and enables new
```

Reviewer: or everything except AdaBoost I think

   use cases that were not possible before.

Both of these projects represent major milestones for `scikit-learn`. In our current
round of priorities, we outline follow-up work for both the website and metadata
routing to further improve and expand on these changes.

#### Mid-term priorities

> From the mid-term priorities, we successfully completed the following:

Reviewer: I feel like this could be restructured to avoid repeating the first sentence here as this:

```
- Short-Term priorities
  - ...
  - ...
- Mid-Term priorities
  - ...
  - ...
 ```

1. Implementation of a new `scikit-learn` estimator, called `TunedThresholdClassifierCV`,
   allowing optimization of the operational decision.
1. Progress on the Array-API adoption by making it possible to train and test a pipeline
   containing a `PCA` and `Ridge` estimator trained on `torch` tensors located on GPU.
1. Some work to improve the developer experience by improving the developer tools
   available in `scikit-learn` and made available for third-party projects.

> #### Long-term priorities

Reviewer: not sure in which section it goes, but we also have done quite a bit with the developer API work.

Author: Indeed. We did quite some work in between the moment I did the retrospective and the updated the new priority.

Author: Actually this is the last point raised in the previous item.

1. Community outreach: Since February 2024, we attended multiple Python conferences
   to promote the latest work on `scikit-learn`. We attended the following conferences:
   PyCon Lithuania, PyCon Italia, CZI Open Science, EuroSciPy, PyData Amsterdam, and
   PyData Paris.

### Computing orchestration

We dedicated resources to maintain the following projects that are related to computing
orchestration: `joblib`, `loky`, `cloudpickle`, `threadpoolctl`, and support for
Python free-threaded mode.

`threadpoolctl` has been released in the past few months and includes better support
for BLAS libraries (FlexBLAS, OpenBLAS, Netlib, Accelerate, etc.).

We dedicated substantial work to ensuring scikit-learn's compatibility with Python 3.13's
free-threaded mode, including extensive testing, necessary adaptations, and reporting
any issues to upstream projects.

### [`fairlearn`](https://github.com/fairlearn/fairlearn)

In the past few months, the main focus of the project has been to ensure that the
estimators developed in `fairlearn` are compatible with `scikit-learn`. To achieve this,
we used the testing framework provided by `scikit-learn` to test the `fairlearn`
estimators.

> 

Reviewer: Could also mention that we also pushed out a release with many compatibility fixes with dependency libraries.

*tamaratanasoska, adrinjalali*

We will contribute to the community roadmap that has recently been updated:
https://docs.google.com/document/d/1rNCqRQP0g6kVplqB4DSW3o9ruXuJQun9yYv6FBouaac/edit?tab=t.0#heading=h.v4tu3p4sz3uj

### [`skops`](https://github.com/skops-dev/skops)

*tamaraatanasoska, adrinjalali*

The main activity in the project relates to non-trivial maintenance tasks
to ensure that the project is in a healthy state and compatible with the latest
versions of `scikit-learn` and `NumPy`.

### [`skrub`](https://github.com/skrub-data/skrub)

*Vincent-Maladiere, glemaitre*

We will help a tackling item from the community roadmap available at:
https://skrub-data.org/stable/vision.html#roadmap

### [`hazardous`](https://github.com/soda-inria/hazardous)

*Vincent-Maladiere, ogrisel, glemaitre*

1. Implement metrics for competing risks
   - Concordance Index
   - Refine metrics API
1. Implement early stopping based on validation score
1. Improve the documentation and examples
1. Start releasing the package
1. Generalize `MultiIncidenceGradientBoosting` as a meta-estimator to wrap any
   classifier that supports `sample_weight`

> ## The maintainers

Reviewer: ```suggestion
## Our contrubotors
```
since not everyone is a maintainer?

# Open source software priorities at Probabl

At Probabl, we continue our dedicated efforts to support and enhance `scikit-learn` and
its ecosystem. In this post, we provide a retrospective on the work we have

> accomplished on `scikit-learn` and other supported open-source packages over the past

Reviewer 2: This gives the impression that the list below was accomplished by probabl alone.

Author: I end up with a section to exactly addressed such impression. So I should move right at the beginning such that it is clear that this is not the case.

Author: I moved the acknowledgement right at the start.

  production.
* Insights from community surveys conducted in 2023 to understand core contributors'
  concerns and priorities. Additionally, we will soon incorporate feedback from the
  ongoing [user `scikit-learn` survey](https://forms.gle/zUXvWjGUN1nWhJ2V6) to further
  align our efforts with user needs and expectations.
* Commitment to revisiting our list of priorities approximately twice a year.

Let's review our achievements and set the stage for our upcoming focus areas.

## Retrospective on the past 6 months

This section summarizes the progress of Probabl's work on the different supported
open-source projects.

### [`scikit-learn`](https://github.com/scikit-learn/scikit-learn)

In a previous [blog
post](https://papers.probabl.ai/scikit-learns-priorities-at-probabl), we outlined
Probabl's priorities for `scikit-learn`. In this section, we report on the progress of
these priorities.

- Short-term priorities

   1. Website Migration: We successfully migrated the `scikit-learn` website to the new
      PyData theme-based website. This modernization effort improves the user
      experience and aligns our documentation with the broader PyData ecosystem.

>    1. SLEP6 - Metadata Routing: We completed adapting most estimators to handle

Reviewer 2: I would either expect "task is completed" or X out of Y remaining estimators where completed in the last 6 months.

Author: I gave more straightforward explanation (only missing `AdaBoost`).

      metadata routing as specified in
      [SLEP6](https://scikit-learn-enhancement-proposals.readthedocs.io/en/latest/slep006/proposal.html).
      This new API consistently handles metadata (e.g., `sample_weight`, `groups`, etc.)
      for almost all estimators. It allows for fewer bugs and enables new
   use cases that were not possible before.

Both of these projects represent major milestones for `scikit-learn`. In our current
round of priorities, we outline follow-up work for both the website and metadata
routing to further improve and expand on these changes.

-  Mid-term priorities

   1. Implementation of a new `scikit-learn` estimator, called

>       `TunedThresholdClassifierCV`, allowing optimization of the operational decision.

Reviewer 2: Is it an estimator or a meta-estimator?
Should `FixedThresholdClassifier` be mentioned, too?

   1. Progress on the Array-API adoption by making it possible to train and test a
      pipeline containing a `PCA` and `Ridge` estimator trained on `torch` tensors
      located on GPU.
   1. Some work to improve the developer experience by improving the developer tools

>       available in `scikit-learn` and made available for third-party projects.

Reviewer 2: For example ...

Author: Done

- Long-term priorities

>    1. General maintenance of the project: We reviewed a substantial number of pull

Reviewer 2: Again, do you have numbers?

Author: I'm not sure if we should specify numbers. Part of the info would be available here (https://blog.scikit-learn.org/resources/). The problem is that I don't know how to split external contributions from contribution from the core team.

Since we have the round-robin cycles, we almost ensure that all issues are triaged in this respect and it increases the hit rate on reviewing PRs.

Would you better like this explanation there or do you think of anyway to report proper numbers.

Author: I added more details regarding what we meant here.

      requests and issues opened by external contributors.

>    1. Python 3.13 support: We worked on supporting Python 3.13 free-threaded mode and

Reviewer 2: Was it together with Nathan from Quansight?

Author: Indeed, it started with this [issue](https://github.com/scikit-learn/scikit-learn/issues/28978) and now Nathan is also involved in this [thread](https://github.com/scikit-learn/scikit-learn/issues/30007) regarding the stress-testing phase. I think Loïc is the person the most involved on this topic from our side.

Do you want me to have an explicit acknowledgement to the involvement of Nathan on this topic?

      set up regular testing to ensure compatibility until it becomes officially
      available.
   1. Releases: Since February 2024, we released `scikit-learn` 1.5.0, 1.5.1, and 1.5.2.
   1. Continuous Integration and Continuous Deployment: We kept our continuous
      integration and continuous deployment infrastructure up to date.
   1. Community outreach: Since February 2024, we attended multiple Python conferences
      to promote the latest work on `scikit-learn`. We attended the following
      conferences: PyCon Lithuania, PyCon Italia, CZI Open Science, EuroSciPy, PyData
      Amsterdam, and PyData Paris.

### Computing orchestration

We dedicated resources to maintain the following projects that are related to computing
orchestration: `joblib`, `loky`, `cloudpickle`, `threadpoolctl`, and support for
Python free-threaded mode.

> `threadpoolctl` has been released in the past few months and includes better support

Reviewer 2: This sounds as if it was released the first time.

Author: What do you mean exactly? The support for FlexiBLAS has only been added lately.

Reviewer 2: What you mean to say is "A new version of theadpoolctl has been released".

> for BLAS libraries (FlexBLAS, OpenBLAS, Netlib, Accelerate, etc.).

Author: ```suggestion
for BLAS libraries (FlexiBLAS, OpenBLAS, Netlib, Accelerate, etc.).
```

[`pyodide`](https://github.com/pyodide/pyodide),
[`jupyterlite`](https://github.com/jupyterlite/jupyterlite). We reported potential
issues upstream and helped run the SciPy test suite for the pyodide project.

### [`hazardous`](https://github.com/soda-inria/hazardous)

The current focus for this project is to define the scope of the library such that it
does not overlap with existing tools. While the code is developed in parallel with a
research project, we are working on improving tests and documentation to make the
library more robust and ready for a first release.

## Focus Areas for the Next 6 Months

As we look ahead to the next six months, we have identified several key areas where we
will concentrate our efforts to further enhance `scikit-learn` and its ecosystem. These
focus areas align with our general objectives and aim to address current challenges
and opportunities in the machine learning landscape.

### [`scikit-learn`](https://github.com/scikit-learn/scikit-learn)

We start by focusing on the future work for `scikit-learn`.

1. Statistical correctness: *ogrisel, snath-xoc, antoinebaker, jeremiedbb*

>    - We will continue our maintenance work to ensure that sample weights are

Reviewer 2: This is not statistical correctness, rather algorithmic/numerical.

Author: Done

      correctly handled in estimators.
   - We will improve the documentation to better explain which cross-validation
      strategies should be used in line with the end-user's goal.

>    - We will work on improving the solvers of linear models.

Reviewer 2: More for my interest: Which ones?

Author: I'll ask @ogrisel to jump in because he would have a clearer idea than me :).

1. New visualizations: *lucyleeow, glemaitre, jeremiedbb*
   - We will develop some additional displays to help with model inspection.
   - We will work on adapting the existing displays to visualize the outputs of
     cross-validation results.
   - We will search for ways to improve the HTML representation of estimators.

> 1. Callbacks API: *jeremiedbb, adrinjalali, glemaitre, ogrisel*

Reviewer 2: I would mention the SLEP.

Author: Done

   - We will work on merging the infrastructure for callbacks together with a callback
     to provide progress bars.
   - Subsequently, we will work on additional callbacks.
1. Improve histogram gradient-boosting: *GaelVaroquaux, glemaitre, lesteve, ogrisel*

>    - We will benchmark the `HistGradientBoosting` estimators to understand the room

Reviewer 2: FYI: My benchmarking indicates that it is mostly the histogram computation.

Author: I should decouple this point to have computation performance and statistical performance on the other side. We plan to look at both: using your benchmark and look at adding the second level parallelism as you mentioned. On the statistical side, Gael reported some benchmark where it seems that `HistGradientBoosting` is doing worse after tuning but this is not clear yet what is the reason. So we would like to have a factual answer to the reason there.

Author: Done

     for improvement of these estimators to match the performance of LightGBM and
     XGBoost.

>    - We will assess if we should merge both `GradientBoosting` and

Reviewer 2: Isn't this a community decision?

Author: Of course. `Assess` here means that someone will report the pros/cons, potential difficulties, by looking in details if we want to go this road. The final call will be based on the consensus.

I will rephrase it because the `we should merge` does not reflect the consensus base decision.

     `HistGradientBoosting` estimators into a single class.
1. Array API: *betatim, ogrisel, lesteve, StefanieSenger, OmarManzoor, EmilyXinyi*
   - We will continue our work on the Array API adoption. Notably, we will work on
     adopting the API for `Nystroem`, `Ridge` and at least one solver of
     `LogisticRegression`.
1. Python 3.13 support: *lesteve, ogrisel*
   - We will continue our work on supporting Python 3.13 free-threaded and release
     wheels and binaries for the new Python version.

> 1. Metadata routing: *adrinjalali, StefanieSenger, glemaitre, OmarManzoor, adam2392*

Reviewer 2: As I did in one of Adrin's PRs, it might be a good idea to contact relecant projects and involve them in the decisions.

Author: I'll mention this as well.

   1. Website Migration: We successfully migrated the `scikit-learn` website to the new
      PyData theme-based website. This modernization effort improves the user
      experience and aligns our documentation with the broader PyData ecosystem.
   1. SLEP6 - Metadata Routing: We completed adapting all estimators (apart from
      `AdaBoost`) to handle metadata routing as specified in
      [SLEP6](https://scikit-learn-enhancement-proposals.readthedocs.io/en/latest/slep006/proposal.html).
      This new API consistently handles metadata (e.g., `sample_weight`, `groups`, etc.)
      for almost all estimators. It allows for fewer bugs and enables new use cases that
      were not possible before.

Both of these projects represent major milestones for `scikit-learn`. In our current
round of priorities, we outline follow-up work for both the website and metadata
routing to further improve and expand on these changes.

-  Mid-term priorities

   1. Implementation of a new `scikit-learn` meta-estimator, called
      `TunedThresholdClassifierCV`, allowing optimization of the operational decision.
      In addition, the `FixedThresholdClassifier` estimator has been added and allows

>       to use a pre-specified threshold for the operational decision.

Reviewer 3: Nitpick, "the operational decision" just sounds a bit odd (not sure why exactly, maybe the 'the' makes it sound like there is a specific decision you are referring to?), does "for operational decision-making" work?

>       to use a pre-specified threshold for the operational decision.

Author: ```suggestion
      to use a pre-specified threshold for the operational decision-making.
```

   1. Progress on the Array-API adoption by making it possible to train and test a
      pipeline containing a `PCA` and `Ridge` estimator trained on `torch` tensors
      located on GPU.
   1. Some work to improve the developer experience by improving the developer tools

>       available in `scikit-learn` and made available for third-party projects. For

Reviewer 3: ```suggestion
      available in `scikit-learn` and making them available for third-party projects. For
```

      instance, we made public some internal APIs that are required to create
      `scikit-learn` estimators (e.g., data validation, estimator tags). Also, we

>       start to improve the tests checking the API conformance of third-party libraries

Reviewer 3: ```suggestion
      started to improve the tests checking the API conformance of third-party libraries
```

      with `scikit-learn`. More work is needed in this area to make it easier to

>       create estimators and check the compatibility with `scikit-learn`.

Reviewer 3: ```suggestion
      create estimators and check their compatibility with `scikit-learn`.
```

- Long-term priorities

   1. General maintenance of the project: We reviewed a substantial number of pull
      requests and issues opened by external contributors.
   1. Python 3.13 support: We worked on supporting Python 3.13 free-threaded mode and
      set up regular testing to ensure compatibility until it becomes officially
      available.
   1. Releases: Since February 2024, we released `scikit-learn` 1.5.0, 1.5.1, and 1.5.2.
   1. Continuous Integration and Continuous Deployment: We kept our continuous
      integration and continuous deployment infrastructure up to date.
   1. Community outreach: Since February 2024, we attended multiple Python conferences

>       to promote the latest work on `scikit-learn`. We attended the following
>       conferences: PyCon Lithuania, PyCon Italia, CZI Open Science, EuroSciPy, PyData
>       Amsterdam, and PyData Paris.

Reviewer 3: Optional suggestion
```suggestion
      to promote the latest work on `scikit-learn`;
      PyCon Lithuania, PyCon Italia, CZI Open Science, EuroSciPy, PyData
      Amsterdam, and PyData Paris.
```

### Computing orchestration

We dedicated resources to maintain the following projects that are related to computing

> orchestration: `joblib`, `loky`, `cloudpickle`, `threadpoolctl`, and support for
> Python free-threaded mode.

Reviewer 3: ```suggestion
orchestration: `joblib`, `loky`, `cloudpickle`, `threadpoolctl`, and worked on supporting
Python free-threaded mode.
```

A new version of `threadpoolctl` was released in the past few months and includes better
support for BLAS libraries (FlexiBLAS, OpenBLAS, Netlib, Accelerate, etc.).

We dedicated substantial work to ensuring scikit-learn's compatibility with Python 3.13's
free-threaded mode, including extensive testing, necessary adaptations, and reporting
any issues to upstream projects.

### [`fairlearn`](https://github.com/fairlearn/fairlearn)

In the past few months, the main focus of the project has been to ensure that the
estimators developed in `fairlearn` are compatible with `scikit-learn`. To achieve this,
we used the testing framework provided by `scikit-learn` to test the `fairlearn`
estimators.

> We also helped at releasing `fairlearn` and make sure it is compatible with the

Reviewer 3: ```suggestion
We also helped with the release of `fairlearn` and made sure that it is compatible with the
```

different upstream dependencies (e.g. `numpy`, `scikit-learn`).

### [`skops`](https://github.com/skops-dev/skops)

> The main activity in the project relates to non-trivial maintenance tasks

Reviewer 3: ```suggestion
The main activity in this project relates to non-trivial maintenance tasks
```

to ensure that the project is in a healthy state and compatible with the latest
versions of `scikit-learn` and `NumPy`.

### [`skrub`](https://github.com/skrub-data/skrub)

In the past few months, we focused on delivering new features such as:

1. Support for `polars` dataframes in `skrub` estimators.
1. Adding interactive reports of dataframes through a `TableReport` that provides
   information to carry out Exploratory Data Analysis (EDA).
1. Easing the creation of predictive models by removing boilerplate code using a factory
   function called `tabular_learner`.
1. Improving the landing page of the `skrub` website to make it more user-friendly and
   informative.

### Python in the browser (WebAssembly)

> We make sure that `scikit-learn` is compatible with the WebAssembly stack:

Reviewer 3: ```suggestion
We made sure that `scikit-learn` is compatible with the WebAssembly stack:
```

   - We will develop some additional displays to help with model inspection.
   - We will work on adapting the existing displays to visualize the outputs of
     cross-validation results.
   - We will search for ways to improve the HTML representation of estimators.
1. [SLEP023](https://github.com/scikit-learn/enhancement_proposals/pull/90) Callbacks
   API: *jeremiedbb, adrinjalali, glemaitre, ogrisel*
   - We will work on merging the infrastructure for callbacks together with a callback
     to provide progress bars.
   - Subsequently, we will work on additional callbacks.
1. Improve histogram gradient-boosting: *GaelVaroquaux, glemaitre, lesteve, ogrisel,
   lorentzenchr*
   - We will benchmark in terms of statistical performance the `HistGradientBoosting`
     estimators to understand the room for improvement of these estimators to match
     the performance of LightGBM and XGBoost.
   - We will work on improving the computational performance following the work of
     `lorentzenchr`.
   - We will assess what are the potential issues, pros, and cons to merge
     `GradientBoosting` and `HistGradientBoosting` estimators into a single class.

>      This will feed the discussion to take a decision on the best way to move forward.

Reviewer 3: ```suggestion
     This will feed into the discussion to make a decision on the best way to move forward.
```

     on test sets.
   - We will work on SLEP11 to provide a unified interface to inspect the "feature
     importances" of estimators.
1. Developer API: *adrinjalali, glemaitre, adam2392, Charlie-XIAO, OmarManzoor*
   - We will continue our work on improving the test suite by migrating some common
     tests to the estimator checks, to finalize common check categories, and to
     improve the associated documentation.
1. Documentation improvements:
   - We plan to further improve the gallery of examples, the contributor guide, and
     the developer guide. To carry on this work, we will organize two dedicated
     sprints focusing especially on the documentation.
1. Project maintenance:
   - We will continue our regular triage of issues and reviewing community pull
     requests.
   - We will make sure to keep our continuous integration and continuous deployment
     infrastructure up to date.
   - We will make sure to reinstate our ASV benchmark allowing use to detect potential
     performance regressions.
1. Releases:
   - We will continue to release on a regular basis.

### Computing orchestration

> We have to adapt all the stack (`joblib`, `loky`, `cloudpickle`, `threadpoolctl`) to the

Reviewer 3: ```suggestion
We have to adapt the whole stack (`joblib`, `loky`, `cloudpickle`, `threadpoolctl`) to the
```

CPython free-threaded mode. This will involve some stress testing to ensure that the
parallelism works as expected.

### [`fairlearn`](https://github.com/fairlearn/fairlearn)

*tamaratanasoska, adrinjalali*

We will contribute to the community roadmap that has recently been updated:
https://docs.google.com/document/d/1rNCqRQP0g6kVplqB4DSW3o9ruXuJQun9yYv6FBouaac/edit?tab=t.0#heading=h.v4tu3p4sz3uj

### [`skops`](https://github.com/skops-dev/skops)

*tamaraatanasoska, adrinjalali*

> The main activity in the project relates to non-trivial maintenance tasks

Reviewer 3: ```suggestion
The main activity in this project relates to non-trivial maintenance tasks
```
